Fixes #9603 - fixed websockets_* set of settings

lzap opened the pull request:
Basically `websockets_encrypt` setting did not work at all. Initially I thought
the only change is not to use certificates when websockets_encrypt is set to
false, but it looks the whole setting handling is incorrect. There are "on" and
"off" switch/case statements which are (no longer?) correct. The setting is
read-only, defined in the settings.yaml and type of boolean.

This fixes it completely.

lzap committed:
Fixes #9603 - fixed websockets_* set of settings

domcleal commented on app/models/setting/provisioning.rb:
```diff
@@ -28,7 +28,7 @@ def self.load_defaults
         self.set('update_ip_from_built_request', N_("Foreman will update the host IP with the IP that made the built request"), false),
         self.set('use_shortname_for_vms', N_("Foreman will use the short hostname instead of the FQDN for creating new virtual machines"), false),
-        self.set('websockets_encrypt', N_("Should Foreman encrypt websockets (VNC console access). Choose on, off or auto."), "auto"),
+        self.set('websockets_encrypt', N_("Should Foreman encrypt websockets (VNC/Spice console access)"), true),
```
"SPICE"

lzap replied:
Fixed.

domcleal commented:
Just a thought while I'm here: you need to migrate existing settings if moving from a string to bool.

lzap commented:
Hmm it used to be in the database but now we deploy it via YAML. I don't get it, what is the history of that?

```
=> #<Setting::Provisioning id: 76, name: "websockets_encrypt", value: nil, description: "Should Foreman encrypt websockets (VNC/SPICE consol...", category: "Setting::Provisioning", settings_type: "string", default: "--- true\n...\n", created_at: "2015-01-20 08:48:17", updated_at: "2015-03-05 14:27:44">
```

lzap commented:
I mean in the UI I see this setting read-only.

domcleal commented:
DB settings can be overridden from config/settings.yaml, and any setting in the config file causes the UI/DB to be read only.  We deploy this in the installer nowadays.

dLobatog commented on lib/ws_proxy.rb:
```diff
@@ -27,8 +27,10 @@ def start_proxy
-      cmd += " --key #{Setting[:websockets_ssl_key]}" if Setting[:websockets_ssl_key]
+      if Setting[:websockets_encrypt]
+        cmd += " --cert #{Setting[:websockets_ssl_cert]}" if Setting[:websockets_ssl_cert]
+        cmd += " --key #{Setting[:websockets_ssl_key]}" if Setting[:websockets_ssl_key]
```
Maybe you want to raise an error if the cert or key are not set, but `websockets_encrypt` is true. Raising the error at `ws_proxy.rb` is probably too late as I can't see how to let the user know about this at this point, probably the controller should handle it instead.

lzap replied:
Fixed.

dLobatog commented on app/controllers/compute_resources_vms_controller.rb:
```diff
@@ -79,14 +79,7 @@ def console
-                 true
-               when 'off'
-                 false
-               else
```
This is what the `auto` setting was supposed to do, however it seems like it would not honor `websockets_encrypt = off` as you pointed out.

dLobatog commented on app/models/setting/provisioning.rb:
```diff
@@ -28,7 +28,7 @@ def self.load_defaults
         self.set('update_ip_from_built_request', N_("Foreman will update the host IP with the IP that made the built request"), false),
         self.set('use_shortname_for_vms', N_("Foreman will use the short hostname instead of the FQDN for creating new virtual machines"), false),
-        self.set('websockets_encrypt', N_("Should Foreman encrypt websockets (VNC console access). Choose on, off or auto."), "auto"),
+        self.set('websockets_encrypt', N_("Should Foreman encrypt websockets (VNC/SPICE console access)"), true),
```
I don't think encrypted websockets works with SPICE, or at least when this was merged we didn't made it work, [#7018](http://projects.theforeman.org/issues/7018)  is open for that reason.

lzap replied:
Changed the string helper.

dLobatog commented:
@lzap Comments inline, also this is missing the migration as @domcleal pointed out, thanks!

lzap commented:
I think the main problem here that this used to be "on" and "off" and "auto" string option but we started to deploy via YAML file as a boolean. So it looks like the best solution here is to migrate to booleans in the database right?

lzap commented:
Ok here is an update. Introduced new simple validation API that is called during rails boot, so we error out early. Introduced new helper exception class that holds bunch of messages, so we can collect all the errors in one run. If you like this approach, I will go ahead and create new button on settings UI screen so users can validate runtime changes and I can also add more validations. Send me your proposals here, or if you prefer I can create new PR for that.

Added migration that converts to boolean.

domcleal commented on db/migrate/20150312144232_migrate_websockets_setting.rb:
```diff
@@ -0,0 +1,16 @@
+  def up
+    old_value = Setting.find_by_name("websockets_encrypt")
+    new_value = (old_value == true || old_value == "true" || old_value == "auto") ? true : false
+    Setting.find_by_name("websockets_encrypt").update_attributes(
```
Tests fail here when migrating a blank database, as the setting may not exist yet.

<pre>
00:03:20.410 undefined method `update_attributes' for nil:NilClass
/var/lib/workspace/workspace/test_develop_pr_core/database/sqlite3/ruby/1.9.3/db/migrate/20150312144232_migrate_websockets_setting.rb:5:in `up'
</pre>

lzap replied:
> Tests fail here when migrating a blank database, as the setting may not exist yet.

Rebased, fixed, good catch. Thanks.

## 

Later,
 Lukas #lzap Zapletal

dLobatog commented:
[test]

domcleal commented:
Every test failed, it doesn't need re-testing.

<pre>
00:03:31.882 + bundle exec rake pkg:generate_source jenkins:unit jenkins:integration TESTOPTS=-v
00:03:45.270 rm -rf jenkins/reports/unit/
00:03:53.872 The Apipie cache is turned off. Enable it and run apipie:cache rake task to speed up API calls.
00:03:54.724 rake aborted!
00:03:54.724 Foreman::MultimessageException: ERF90-6855 [Foreman::MultimessageException]: The following settings are invalid: [["missing required option websockets_ssl_key", "missing required option websockets_ssl_cert"]]
00:03:54.725 /var/lib/workspace/workspace/test_develop_pr_core/database/postgresql/ruby/1.9.3/app/services/foreman/multimessage_exception.rb:27:in `raise_if_any'
00:03:54.725 /var/lib/workspace/workspace/test_develop_pr_core/database/postgresql/ruby/1.9.3/app/models/setting.rb:173:in `validate_all'
00:03:54.725 /var/lib/workspace/workspace/test_develop_pr_core/database/postgresql/ruby/1.9.3/config/initializers/foreman.rb:22:in `<top (required)>'
</pre>

lzap commented:
I haven't realized that we now check for certificate files. Therefore webockets SSL is turned off by default.

lzap commented:
Before you say "hey this is not secure", please keep in mind our installer usually sets cert/key and explicitly enable SSL mode. But we need some sane defaults for development mode.

domcleal commented on app/services/foreman/multimessage_exception.rb:
```diff
@@ -0,0 +1,30 @@
+      if @messages.size > 0
+        super + ": #{@messages.split(', ')}"
+      else
+        super + ": none"
```
missing i18n?

domcleal commented on app/services/foreman/multimessage_exception.rb:
```diff
@@ -0,0 +1,30 @@
+
+    def message
+      if @messages.size > 0
+        super + ": #{@messages.split(', ')}"
```
I haven't followed precisely where this is used, but the colon should be part of the message preceding it for translation purposes.

lzap replied:
Validator stacks error messages here, similarly to `@errors` array.

Fixed both thanks.

domcleal commented:
Oh, success!  The upgrade tests caught a bug I think.

<pre>
00:04:29.905 + bundle exec rake db:migrate
00:04:39.464 The Apipie cache is turned off. Enable it and run apipie:cache rake task to speed up API calls.
00:04:39.855 [deprecated] I18n.enforce_available_locales will default to true in the future. If you really want to skip validation of your locale you can set I18n.enforce_available_locales = false to avoid this message.
00:04:39.982 rake aborted!
00:04:39.983 Foreman::MultimessageException: ERF90-6855 [Foreman::MultimessageException]: The following settings are invalid: [["missing required option websockets_ssl_key", "missing required option websockets_ssl_cert"]]
00:04:39.983 /var/lib/workspace/workspace/test_branch_upgrade/database/postgresql/ruby/2.0.0/app/services/foreman/multimessage_exception.rb:27:in `raise_if_any'
00:04:39.983 /var/lib/workspace/workspace/test_branch_upgrade/database/postgresql/ruby/2.0.0/app/models/setting.rb:173:in `validate_all'
00:04:39.983 /var/lib/workspace/workspace/test_branch_upgrade/database/postgresql/ruby/2.0.0/config/initializers/foreman.rb:22:in `<top (required)>'
</pre>

To be clear what they do - they checkout 1.7-stable or 1.8-stable, run db:migrate & db:seed, then check out your PR (merged onto develop) and run db:migrate again, like a user upgrading.  That last db:migrate is failing with the above error.

domcleal commented on app/models/setting/provisioning.rb:
```diff
@@ -28,7 +28,7 @@ def self.load_defaults
         self.set('update_ip_from_built_request', N_("Foreman will update the host IP with the IP that made the built request"), false),
         self.set('use_shortname_for_vms', N_("Foreman will use the short hostname instead of the FQDN for creating new virtual machines"), false),
-        self.set('websockets_encrypt', N_("Should Foreman encrypt websockets (VNC console access). Choose on, off or auto."), "auto"),
+        self.set('websockets_encrypt', N_("Should Foreman encrypt websockets (VNC console access)"), false),
```
Perhaps the value of SETTINGS[:require_ssl] would be a more accurate default?  (We use this for protocol selection in foreman_url already.)

If websockets_encrypt is disabled when SSL's enabled, then you get a mismatch and secure_headers prevents it loading.

lzap replied:
Took the default.

lzap commented:
Kudos to upgrade tests, but it looks like after upgrade we leave `websockets_encrypt` turned on but key/cert pair is empty. I am not sure what I can do with this. Maybe to disable the `websockets_encrypt` when one of these is missing and just send a warning to the logs?

lzap commented:
Ok this looks like the most sane approach.

```
  def self.validate_settings(errors)
    if Setting["websockets_ssl_key"].empty? || Setting["websockets_ssl_cert"].empty?
      Rails.logger.warn("Either websockets_ssl_key or websockets_ssl_cert is missing, turning off websockets_encrypt option.")
      Setting["websockets_encrypt"] = false
    end
  end
```

lzap commented:
So we _permanently_ turn off websockets_encrypt if key or cert is missing. It would not work anyway. On production installs this will not happen by default as we deploy the key/cert pair. This can happen on development installs.

domcleal commented on db/migrate/20150312144232_migrate_websockets_setting.rb:
```diff
@@ -0,0 +1,16 @@
+class MigrateWebsocketsSetting < ActiveRecord::Migration
+  def up
+    return unless old_value = Setting.find_by_name("websockets_encrypt")
+    new_value = (old_value == true || old_value == "true" || old_value == "auto") ? true : false
```
`old_value` is a Setting isn't it, not the value of the setting?  I don't understand why you're comparing a Setting to true etc, because as far as I can see, it'll never be true.

domcleal commented on config/initializers/foreman.rb:
```diff
@@ -19,6 +19,7 @@
   end
 
   Setting.descendants.each(&:load_defaults)
+  Setting.validate_all
```
This will work OK on startup, but how does validation work from the settings controllers?  If I change the value of these things, there's no validation performed until the next time the server starts, which is going to be problematic.

Validation makes most sense at the model level I think, but validating at any time over multiple settings is going to cause problems for users who can only change one thing at a time.

lzap replied:
You answered your concern basically, I don't want to iterate through all of them. Seems this is broken by design. Will try different approach here.

domcleal commented on app/models/setting/general.rb:
```diff
@@ -28,4 +28,10 @@ def self.load_defaults
   end
+
+  def self.validate_settings(errors)
+    ["idle_timeout", "proxy_request_timeout", "entries_per_page"].each do |s|
```
This is unrelated, should be in its own PR... but it also should be using the regular model validations.  This validate_settings thing will be confusing if we use it for both validation of multiple inter-dependent settings and validation of standalone settings, which we already do today in models.

lzap replied:
Yeah, I will remove and try to do per-entry approach. But I will keep `entries_per_page` as a harmless example for the general settings just to have something in the file.

lzap commented:
Ok how about this, simple per-record validation "framework". Unfortunately in our case we cannot error out, so we force-reset the boolean with a warning message. I can add two other validators for key/cert in case user first set them and then empty one of them.

domcleal commented on app/models/setting/provisioning.rb:
```diff
@@ -20,9 +20,20 @@ def self.load_defaults
         self.set('libvirt_default_console_address', N_("The IP address that should be used for the console listen address when provisioning new virtual machines via Libvirt"), "0.0.0.0"),
         self.set('update_ip_from_built_request', N_("Foreman will update the host IP with the IP that made the built request"), false),
         self.set('use_shortname_for_vms', N_("Foreman will use the short hostname instead of the FQDN for creating new virtual machines"), false),
+        self.set('websockets_encrypt', N_("VNC/SPICE websocket proxy console access encryption (websockets_ssl_key/cert setting required)"), SETTINGS[:require_ssl]),
```
These duplicate auth.rb, no?

lzap replied:
Sorry what?

domcleal replied:
auth.rb in the same directory:

<pre>
14:00 $ grep websockets_ app/models/setting/*
app/models/setting/auth.rb:        self.set('websockets_encrypt', N_("Should Foreman encrypt websockets (VNC console access). Choose on, off or auto."), "auto"),
app/models/setting/auth.rb:        self.set('websockets_ssl_key', N_("Private key that Foreman will use to encrypt websockets "), nil),
app/models/setting/auth.rb:        self.set('websockets_ssl_cert', N_("Certificate that Foreman will use to encrypt websockets "), nil),
app/models/setting/provisioning.rb:        self.set('websockets_encrypt', N_("VNC/SPICE websocket proxy console access encryption (websockets_ssl_key/cert setting required)"), SETTINGS[:require_ssl]),
app/models/setting/provisioning.rb:        self.set('websockets_ssl_key', N_("Private key that Foreman will use to encrypt websockets "), nil),
app/models/setting/provisioning.rb:        self.set('websockets_ssl_cert', N_("Certificate that Foreman will use to encrypt websockets "), nil)
</pre>


Probably a bad rebase over fb293de.

lzap replied:
Oh that explains it! Nasty. Sure thing.

domcleal commented on app/models/setting/provisioning.rb:
```diff
@@ -20,9 +20,20 @@ def self.load_defaults
+  def validate_websockets_encrypt(record)
+    if Setting["websockets_ssl_key"].empty? || Setting["websockets_ssl_cert"].empty?
+      # we cannot fail due to compatibility so we force set it here
+      Rails.logger.warn("Either websockets_ssl_key or websockets_ssl_cert is missing, turning off websockets_encrypt option.")
```
Is there a good reason we can't just record a model error in validation, instead of logging and silently changing it?

lzap replied:
I was thinking migration, but I should be able to workaround it I guess. During migration, if key/cert is missing and encrypt is on, I can turn it off apparently.

domcleal commented:
Please fix the test failure.

lzap commented:
This indeed works! Nice.

``` diff
diff --git a/app/models/setting/provisioning.rb b/app/models/setting/provisioning.rb
index 3169dc0..aa6ba75 100644
--- a/app/models/setting/provisioning.rb
+++ b/app/models/setting/provisioning.rb
@@ -28,12 +28,22 @@ class Setting::Provisioning < Setting

     true
   end
-  
+
   def validate_websockets_encrypt(record)
     if Setting["websockets_ssl_key"].empty? || Setting["websockets_ssl_cert"].empty?
-      # we cannot fail due to compatibility so we force set it here
-      Rails.logger.warn("Either websockets_ssl_key or websockets_ssl_cert is missing, turning off websockets_encrypt option.")
-      record.value = false
+      record.errors[:base] << _("Unable to turn on websockets_encrypt, either websockets_ssl_key or websockets_ssl_cert is missing")
+    end
+  end
+
+  def validate_websockets_ssl_key(record)
+    if record.value.empty? && Setting["websockets_encrypt"]
+      record.errors[:base] << _("Unable to unset websockets_ssl_key when websockets_encrypt is on")
+    end
+  end
+
+  def validate_websockets_ssl_cert(record)
+    if record.value.empty? && Setting["websockets_encrypt"]
+      record.errors[:base] << _("Unable to unset websockets_ssl_cert when websockets_encrypt is on")
     end
   end
 end
diff --git a/db/migrate/20150312144232_migrate_websockets_setting.rb b/db/migrate/20150312144232_migrate_websockets_setting.rb
index c23a99c..160ad13 100644
--- a/db/migrate/20150312144232_migrate_websockets_setting.rb
+++ b/db/migrate/20150312144232_migrate_websockets_setting.rb
@@ -1,12 +1,10 @@
 class MigrateWebsocketsSetting < ActiveRecord::Migration
   def up
-    return unless old = Setting.find_by_name("websockets_encrypt")
-    new_value = (old.value == true || old.value == "true" || old.value == "auto") ? true : false
-    Setting.find_by_name("websockets_encrypt").update_attributes(
-      :settings_type => "boolean",
-      :value => new_value,
-      :default => true
-    )
+    return unless encrypt = Setting.find_by_name("websockets_encrypt")
+    encrypt.settings_type = "boolean"
+    encrypt.value = (encrypt.value == true || encrypt.value == "true" || encrypt.value == "auto") ? true : false
+    encrypt.default = true
+    encrypt.save(:validate => false)
   end

   def down
```

lzap commented:
One additional change - I changed the migration to set the default value to `false` actually. Makes sense since there are no keys/certs by default. We override this in our installer.

domcleal commented:
You need to fix the duplication between auth.rb/provisioning.rb else this is going to work unpredictably.

lzap commented:
Moved to the right place!

lzap commented:
Not so sure if you like this, but turned off validations for tests... ;-)

domcleal commented:
Leaving as WoC until tests pass.

lzap commented:
Can't reproduce on my laptop, triggering new CI run with a small parch as per IRC.

lzap commented:
You were right, the validation was not testing it properly. Fixed.

domcleal commented on app/models/setting.rb:
```diff
@@ -18,6 +18,12 @@ def validate_each(record, attribute, value)
 
+  class ValueValidator < ActiveModel::Validator
+    def validate(record)
+      record.send("validate_#{record.name}", record) unless Rails.env.test?
```
Why the Rails.env conditional?

lzap replied:
> > -      record.send("validate_#{record.name}", record) unless Rails.env.test?
> 
> Why the Rails.env test?

Leftover, amended.

## 

Later,
 Lukas #lzap Zapletal

domcleal commented on db/migrate/20150312144232_migrate_websockets_setting.rb:
```diff
@@ -0,0 +1,14 @@
+  def up
+    return unless encrypt = Setting.find_by_name("websockets_encrypt")
+    encrypt.settings_type = "boolean"
+    encrypt.value = (encrypt.value == true || encrypt.value == "true" || encrypt.value == "auto") ? true : false
```
When the key + cert are empty, while websockets_encrypt is "auto", this changes the new value to true.  Should it check if the key + cert are defined and if not, set it to false?

I was surprised to see my auto/unset/unset config change to true/unset/unset when it's really false.

lzap replied:
> I was surprised to see my auto/unset/unset config change to true/unset/unset when it's really false.

Good catch, fixed.

## 

Later,
 Lukas #lzap Zapletal

domcleal commented on db/migrate/20150312144232_migrate_websockets_setting.rb:
```diff
@@ -0,0 +1,19 @@
+    return unless encrypt = Setting.find_by_name("websockets_encrypt")
+    encrypt.settings_type = "boolean"
+    if encrypt.value == "auto"
+      encrypt.value = (Setting.find_by_name("websockets_ssl_key").nil? ||
```
Check the `.value` of the settings, or use `Setting[:websockets_ssl_key]` to see the combined (default+value) value.  It's currently checking if a Setting object is nil or not.

domcleal commented on db/migrate/20150312144232_migrate_websockets_setting.rb:
```diff
@@ -0,0 +1,19 @@
+      encrypt.value = (Setting.find_by_name("websockets_ssl_key").nil? ||
+                       Setting.find_by_name("websockets_ssl_cert").nil?) ? false : true
+    else
+      encrypt.value = (encrypt.value == true || encrypt.value == "true") ? true : false
```
Maybe use Foreman::Cast.to_bool

lzap replied:
> Maybe use Foreman::Cast.to_bool

Amended.

## 

Later,
 Lukas #lzap Zapletal

domcleal commented on db/migrate/20150312144232_migrate_websockets_setting.rb:
```diff
@@ -0,0 +1,19 @@
+    return unless encrypt = Setting.find_by_name("websockets_encrypt")
+    encrypt.settings_type = "boolean"
+    if encrypt.value == "auto"
+      encrypt.value = (Setting[:websockets_ssl_key].nil? ||
```
Nitpick, the old test was for `.blank?` rather than `.nil?`, since a user might set it to an empty value.

lzap replied:
> Nitpick, the old test was for `.blank?` rather than `.nil?`, since a user might set it to an empty value.

Empty sets nil, but sure, that's better.

## 

Later,
 Lukas #lzap Zapletal

domcleal commented:
Merged as 97c3447 for 1.10, thanks @lzap.
